Userland/Libraries/LibGUI/EditingEngine.cpp

```diff
@@ -97,10 +111,24 @@ bool EditingEngine::on_key(KeyEvent const& event)
         if (m_editor->cursor().line() < (m_editor->line_count() - 1) || m_editor->is_wrapping_enabled()) {
             m_editor->update_selection(event.shift());
         }
-        move_one_down(event);
-        if (event.shift() && m_editor->selection().start().is_valid()) {
-            m_editor->selection().set_end(m_editor->cursor());
-            m_editor->did_update_selection();
+        auto selection = m_editor->selection();
+        auto const cursor_column = m_editor->cursor().column();
+        if (move_one_down(event) == DidMoveALine::Yes) {
+            if (!selection.is_valid()) {
+                m_editor->set_cursor(m_editor->cursor().line(), cursor_column);
+            } else {
+                auto const final_start_line = selection.start().line() + 1;
+                auto const final_end_line = selection.end().line() + 1;
+                selection.start().set_line(final_start_line);
+                selection.end().set_line(final_end_line);
+                m_editor->set_selection(selection);
+                m_editor->did_update_selection();
+            }
+        } else {
+            if (event.shift() && m_editor->selection().start().is_valid()) {
+                m_editor->selection().set_end(m_editor->cursor());
+                m_editor->did_update_selection();
+            }
```

Comment:
it looks like you duplicated a lot of logic here. could this be a lambda that takes in the direction the line was moved?

Reply:
Even more code can be merged with functions that take the direction. I'm currently working on making a `MoveLineUpOrDownCommand` to take advantages of commands, and I also ended deleting up and down functions. 

I planned to make on other PR to split logical changes and code refactoring to ease the review process. But I can definitely merge them.